prelude/src/core/Result.sk

```diff
@@ -28,6 +28,10 @@ base class Result<+T, +E> {
   | Failure(error: E)
   | Success(value: T)
 
+  fun toResult<U>(): Result<T, Failure<U, E>>
+  | Success(val) -> Success(val)
+  | Failure(err) -> Failure(Failure(err))
```

Comment:
I hate to go back and forth, but when thinking of the `toResult` name I was expecting that this function would be the identity rather than lifting the `Failure` case. That plus the comment about not having good error messages when implementing `?` as a syntactic expansion makes me think that the `toResult` name isn't great. Considering these functions by themselves, they are essentially lifting the failure case into the "failure monad", so `liftFailure` would make sense. But maybe there is a different name that would make an error message coming from using `e?` when `e` doesn't implement the needed function more understandable. With no keyword in `e?` though, it is hard to guess what name would be a good guide. I guess we could add a special case in the compiler where undefined function error messages originate that checks if the needed function is this one and if so adjusts the message to include something like "e? can only be used if the type of e implements liftFailure<U>(): Result<T, _>, such as Option or Result".

Reply:
I agree that the `toResult()` name is weird. I'll update the name to `liftFailure()`.

Since we're now expanding the `?` operator when building the AST from the CST, I don't think there is a solid way to customize error messages: if for some reason the user defines a `liftFailure()` method on a type, but calls it on the wrong type, they would get an error about the `?` operator although it has nothing to do with it.

Reply:
Yes, making the error reporting code look for `liftFailure` could lead to mis-identifying something as a mistake related to `?`. But it might be worth it anyhow if the error message was worded carefully to indicate that `?` is only one possible explanation. All of the "did you mean" stuff, in skip and other languages, also falls into that category.